deleting ghstack id from commit message to force ghstack to create a new PR. last PR metadata was incorrect

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44197 deleting ghstack id from commit message to force ghstack to create a new PR. last PR metadata was incorrect**

Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

adding a beta parameter to the smooth_l1 loss fn

bdhirsh committed:
adding a beta parameter to the smooth_l1 loss fn

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"




[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/PointwiseOps.h:
```diff
@@ -11,10 +11,11 @@ struct TensorIterator;
 namespace native {
 
 using pointwise_fn = void (*)(TensorIterator&, Scalar scalar);
+using pointwise_fn_with_beta = void (*)(TensorIterator&, Scalar scalar, Scalar beta);
```
lmk if you think the name should be more general. Right now smooth_l1_backwards looks like the only point wise op here that takes in two scalar params

bdhirsh commented on aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp:
```diff
@@ -46,28 +46,31 @@ static void addcdiv_cpu_kernel(TensorIterator& iter, Scalar value) {
           return norm_val * grad_output;
         else
-          return norm_val * x * grad_output;
+          return norm_val * x * grad_output / beta_val;
```
the smooth l1 loss when |X| < 1 now divides by the beta parameter, so the derivative does as well.

bdhirsh commented on aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp:
```diff
@@ -46,28 +46,31 @@ static void addcdiv_cpu_kernel(TensorIterator& iter, Scalar value) {
+      [norm_val_vec, beta_val_vec, neg_1_vec, pos_1_vec](
          Vec256<scalar_t> input, Vec256<scalar_t> target, Vec256<scalar_t> grad_output) -> Vec256<scalar_t> {
         auto x = input - target;
+        x /= beta_val_vec;
```
performance question:

I kept the clamping logic instead of using conditionals just to minimize the amount of change that I introduce. But that comes with the downside of always having to divide by beta_val_vec. Versus the conditional code would look something like (only have to do the division in the last case):
`if (x > beta_val_vec) return pos_1_vec else if (x < -beta_val_vec) return neg_1_vec else return norm_val_vec * x * grad_output / beta_val_vec`

How carefully should we be looking at performance for a change like this?

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"




[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6042,23 +6042,23 @@
     CUDA: legacy::cuda::_thnn_nll_loss2d_backward
 
-- func: smooth_l1_loss.out(Tensor self, Tensor target, int reduction=Mean, *, Tensor(a!) out) -> Tensor(a!)
+- func: smooth_l1_loss.out(Tensor self, Tensor target, int reduction=Mean, float beta=1.0, *, Tensor(a!) out) -> Tensor(a!)
```
I'm a little confused by the use of floats vs. doubles in native_functions, and how that trickles through everywhere else.

I originally set `beta` as a double, but that caused the codegen portion of the build process to complain (codegen looks like it only accepts a subset of primitives).

I tried to set `beta` as a float in all other function signatures in the other files, but I'm still running into issues. Currently, beta is set as a double in `autocast_mode.cpp` and `FunctionsManual.h`. I get linker errors in each of the test files of the form `undefined reference to 'at::native::smooth_l1_loss(at::Tensor const&, at::Tensor const&, long, double)'`.

When I try changing `beta` to a float in those files to match the other definition, I get compiler errors of the form 
 `error: could not convert template argument ‘at::smooth_l1_loss’ from ‘at::Tensor (*)(..., **double**) {aka at::Tensor (*)(..., **double**)}’ to ‘at::Tensor (*)(..., **float**)’`

I don't think I'm following the codegen workflow for how these files integrate with each other.

albanD reviewed (commented):
Hey,

Quick question about typing: What should the type of beta be?
It seems to be float in the native_functions level. But it is a Scalar in the kernels. And it is a double in the backward formulas.
Should it be a Scalar all the time to have full precision all the time?
I think the backward and native_functions should at least agree?

bdhirsh commented:
> Hey,
> 
> Quick question about typing: What should the type of beta be?
> It seems to be float in the native_functions level. But it is a Scalar in the kernels. And it is a double in the backward formulas.
> Should it be a Scalar all the time to have full precision all the time?
> I think the backward and native_functions should at least agree?

I talked to Greg about this briefly and he cleared most of it up. I'm going to make it a double rather than a Scalar, since not all of the types that Scalar can take on make sense for the beta param. e.g. We probably don't want it to be a complex, since beta is used by comparison operators

bdhirsh committed:
Update on "deleting ghstack id from commit message to force ghstack to create a new PR. last PR metadata was incorrect"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

adding a beta parameter to the smooth_l1 loss fn

[ghstack-poisoned]
